app/code/Magento/ImportExport/Model/Import/Source/FileParser/CompositeParserFactory.php

```diff
@@ -0,0 +1,62 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\ImportExport\Model\Import\Source\FileParser;
+
+/**
+ * File Parser Composite
+ *
+ */
+class CompositeParserFactory implements ParserFactoryInterface
+{
+    /** @var ParserFactoryInterface[] */
+    private $parserFactories;
+
+    public function __construct(array $parserFactories = [])
+    {
+        $this->parserFactories = [];
+
+        foreach ($parserFactories as $parserFactory) {
+            $this->addParserFactory($parserFactory);
+        }
+    }
+
+    public function create($path, array $options = [])
+    {
+        foreach ($this->parserFactories as $parserFactory) {
+            $parser = $this->createParserIfSupported($parserFactory, $path, $options);
+
+            if ($parser === false) {
+                continue;
+            }
+
+            return $parser;
+        }
+
+        $this->thereWasNoParserFound($path);
+    }
+
+    public function addParserFactory(ParserFactoryInterface $parserFactory)
+    {
+        $this->parserFactories[] = $parserFactory;
+    }
+
+    private function thereWasNoParserFound($path)
+    {
+        throw new UnsupportedPathException($path);
+    }
+
+    private function createParserIfSupported(ParserFactoryInterface $parserFactory, $path, array $options)
+    {
+        try {
+            $parser = $parserFactory->create($path, $options);
+        } catch (UnsupportedPathException $e) {
+            return false;
```

Comment:
It is a valid case for actual CSV file parser factory or for ZIP file parser factory as if they are used directly by a client, returning false is a terrible thing. Adding if ($status) is worse practice as it increases cyclomatic complexity of client code and hidden bugs for unchecked conditions, exceptions look better in this case in my opinion.

Also, a CompositeFactory does throw the same exception if no other parser factories gave a valid parser instance, so unexpected path gets notified up the chain. 

We cannot use here common for Magento 2 "file extension based" strategy pattern, as file path might be a stream or even a temporary file without the extension, so offloading logic to parser factory looks more clean and extensible. As well factory always throws UnsupportedPathException if the file cannot be read by any of the parsers. 

If you have a better idea how to solve this issue, I would be happy to change the code to that structure.